avoid concrete examples for missingResourceError

CaoShuFeng opened the pull request:
missingResourceError uses pod and services as an example in error message. However some sub-commands doesn't support pod/service, this change use `<resource> <name>` instead of concrete examples.
Before this change:
```console
$ kubectl set subject
error: You must provide one or more resources by argument or filename.
Example resource specifications include:
   '-f rsrc.yaml'
   '--filename=rsrc.json'
   'pods my-pod'
   'services'
```

After this change:
```console
$ kubectl set subject
error: You must provide one or more resources by argument or filename.
Example resource specifications include:
   '-f rsrc.yaml'
   '--filename=rsrc.json'
   '<resource> <name>'
   '<resource>'
```

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
avoid concrete examples for missingResourceError
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/45582)
<!-- Reviewable:end -->

rootfs commented:
/assign @janetkuo

janetkuo commented:
@k8s-bot ok to test

janetkuo commented:
@CaoShuFeng, thanks for the fix. IMO it's a valid error when the resource isn't specified (`missingResourceError`). It's one of general `kubectl` usage errors, so I don't want to special case it for `kubectl set`. 

I'd like to hear more on why the error message is odd. What part of the original error message is confusing?

janetkuo reviewed (commented):
@CaoShuFeng, thanks for the fix. IMO it's a valid error when the resource isn't specified (`missingResourceError`). It's one of general `kubectl` usage errors, so I don't want to special case it for `kubectl set`. 

I'd like to hear more on why you think the error message is odd. What part of the original error message is confusing?

CaoShuFeng commented:
@janetkuo 
For example, "kubectl set subject" only support "rolebinding/clusterrolebinding" but the help message uses "pods my-pod/services" as an example. In fact, we can't use "kubectl set subject pods".

CaoShuFeng commented:
@k8s-bot verify test this

janetkuo commented:
@CaoShuFeng I see, that makes sense. Please put the options validation code in `Validate`, ref kubectl convention https://github.com/kubernetes/community/blob/master/contributors/devel/kubectl-conventions.md#command-implementation-conventions

CaoShuFeng commented:
@janetkuo 
Looks like someone else met a similar problem before:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/resource/builder.go#L91
So is use this API to check the error.

CaoShuFeng commented:
> Please put the options validation code in Validate

I tried your suggestion, but I found that error happens before we call Validate(), so it's default to check options in Validate().

CaoShuFeng commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

janetkuo commented:
@CaoShuFeng I see, the `missingResourceError` is returned by resource builder. We can just modify the example part of `missingResourceError` -- don't use concrete examples since it's a generic error for kubectl, but not all commands support the resources specified in the example. 

In summary, change it to something like:
```go
var missingResourceError = fmt.Errorf(`You must provide one or more resources by argument or filename.
Example resource specifications include:
   '-f app.yaml'
   '--filename=app.json'
   '<resource> <name>'
   '<resource>'`)
```

CaoShuFeng committed:
modify the example part of missingResourceError

"missingResourceError" is a generic error for kubectl, so we should
not use concrete examples in the error message.

CaoShuFeng commented:
@janetkuo 
Done.@

CaoShuFeng commented:
@k8s-bot unit test this

janetkuo commented:
Thanks!
/lgtm
/approve

CaoShuFeng commented:
@k8s-bot gce etcd3 e2e test this
